examples/pettingzoo/meltingpot_env.py

```diff
@@ -0,0 +1,149 @@
+from ml_collections import config_dict
+from gym import spaces
+import numpy as np
+from functools import lru_cache
+
+from gym.utils import EzPickle
+from pettingzoo.utils import wrappers
+from pettingzoo.utils.conversions import parallel_to_aec_wrapper
+from pettingzoo.utils.env import ParallelEnv
+import tree
+import matplotlib.pyplot as plt
+
+PLAYER_STR_FORMAT = 'player_{index}'
+MAX_CYCLES = 1000
+
+def _timestep_to_observations(timestep: dm_env.TimeStep):
+  gym_observations = {}
+  for index, observation in enumerate(timestep.observation):
+    gym_observations[PLAYER_STR_FORMAT.format(index=index)] = {
+        key: value for key, value in observation.items() if 'WORLD' not in key
+    }
+  return gym_observations
+
+
+def _remove_world_observations_from_space(
+    observation: spaces.Dict) -> spaces.Dict:
+  return spaces.Dict({
+      key: observation[key] for key in observation if 'WORLD' not in key})
+
+
+def _spec_to_space(spec: tree.Structure[dm_env.specs.Array]) -> spaces.Space:
+  """Converts a dm_env nested structure of specs to a Gym Space.
+
+  BoundedArray is converted to Box Gym spaces. DiscreteArray is converted to
+  Discrete Gym spaces. Using Tuple and Dict spaces recursively as needed.
+
+  Args:
+    spec: The nested structure of specs
+
+  Returns:
+    The Gym space corresponding to the given spec.
+  """
+  if isinstance(spec, dm_env.specs.DiscreteArray):
+    return spaces.Discrete(spec.num_values)
+  elif isinstance(spec, dm_env.specs.BoundedArray):
+    return spaces.Box(spec.minimum, spec.maximum, spec.shape, spec.dtype)
+  elif isinstance(spec, dm_env.specs.Array):
+    if np.issubdtype(spec.dtype, np.floating):
+      return spaces.Box(-np.inf, np.inf, spec.shape, spec.dtype)
+    elif np.issubdtype(spec.dtype, np.integer):
+      info = np.iinfo(spec.dtype)
+      return spaces.Box(info.min, info.max, spec.shape, spec.dtype)
+    else:
+      raise NotImplementedError(f'Unsupported dtype {spec.dtype}')
+  elif isinstance(spec, (list, tuple)):
+    return spaces.Tuple([_spec_to_space(s) for s in spec])
+  elif isinstance(spec, dict):
+    return spaces.Dict({key: _spec_to_space(s) for key, s in spec.items()})
+  else:
+    raise ValueError('Unexpected spec: {}'.format(spec))
+
+def parallel_env(env_config, max_cycles=MAX_CYCLES):
+    return _parallel_env(env_config, max_cycles)
+
+def raw_env(env_config, max_cycles=MAX_CYCLES):
+    return parallel_to_aec_wrapper(parallel_env(env_config, max_cycles))
+
+def env(env_config, max_cycles=MAX_CYCLES):
+    aec_env = raw_env(env_config, max_cycles)
+    aec_env = wrappers.AssertOutOfBoundsWrapper(aec_env)
+    aec_env = wrappers.OrderEnforcingWrapper(aec_env)
+    return aec_env
+
+class MeltingPotPettingZooEnv(ParallelEnv):
+  """An adapter between the Melting Pot substrates and PettingZoo's ParallelEnv"""
+
+  def __init__(self, env_config, max_cycles):
+    self.env_config = config_dict.ConfigDict(env_config)
+    self.max_cycles = max_cycles
```

Comment:
Does this need to be public? Is it part of the PettingZoo API?

Reply:
No, the PettingZoo public API mainly consists of the three methods `parallel_env`, `env`, and `raw_env`.

The ParallelEnv API (returned by `parallel_env`) steps all agents synchronously and is used in the example.

The `raw_env` returns an AEC environment ([Docs](https://www.pettingzoo.ml/api), [Paper](https://arxiv.org/abs/2009.14471)), which steps each agent individually.

The `env` function is `raw_env` with some thin wrappers to check that input actions are in the action space, etc.

I can make `MeltingPotPettingZooEnv` private if you'd prefer; PettingZoo recommends that you use the three functions instead of the raw environment class to initialize the environment. The _ParallelEnv class is a convenience to make `EzPickle` work.

Reply:
Thanks for the context. It seems then that this module really should only have 3 public members: `env`, `raw_env` and `parallel_env`. Please make everything else private. And likewise, make the all the members of classes (that make sense to do so) private.